### PR TITLE
Stop using mypy_extensions

### PR DESCRIPTION
**Pull request by wsanchez**
Remove some unused type extensions, fixing #414.

This also eliminates use of `mypy_extensions`, fixing #415, as those types were the users of it.

### CODE REVIEW

File: src/klein/_app.py [hunk 1]
```diff
@@ -39,17 +39,41 @@
...
-    [Optional["Klein"], IRequest, Failure], KleinRenderable
-]
+
+
+# Protocol is new in Python 3.8
+try:
+    from typing import Protocol
+except ImportError:
+    KleinRoute = Callable[..., KleinRenderable]
+    KleinErrorHandler = Callable[..., KleinRenderable]
```
Comment (wsanchez) [inline src/klein/_app.py:55]: This is wimpier than the prior spec but when we do type checking, we use Python 3.8+, so the Protocol below.

File: src/klein/_form.py [hunk 2]
```diff
@@ -451,13 +451,6 @@ def defaultValidationFailureHandler(
 
 
 _requirerFunctionWithForm = Any
-_routeCallable = Any
-_routeDecorator = Callable[
-    [_routeCallable, DefaultNamedArg(Any, "__session__")], _routeCallable
-]
-_validationFailureHandler = Callable[
-    [Optional[object], IRequest, "Form", Dict[str, str]], Element
-]
```
Comment (wsanchez) [inline src/klein/_form.py:460]: These are unused ballast

File: src/klein/_session.py [hunk 3]
```diff
@@ -199,20 +198,6 @@ def procureSession(
...
-_kleinRenderable = Any
-_routeCallable = Any
-_kleinCallable = Callable[..., _kleinRenderable]
-_kleinDecorator = Callable[[_kleinCallable], _kleinCallable]
-_requirerResult = Callable[
-    [Arg(_routeCallable, "route"), KwArg(Any)],
-    Callable[[_kleinCallable], _kleinCallable],
-]
-
-
```
Comment (wsanchez) [inline src/klein/_session.py:215]: These are unused ballast

File: src/klein/_typing.py [hunk 4]
```diff
@@ -1,28 +1,13 @@
...
-
-    def DefaultNamedArg(*ignore):
-        pass
 
-    NoReturn = None
+else:
 
-    ifmethod = _ifmethod
+    def ifmethod(method: Callable) -> Callable:
+        return method
```
Comment (wsanchez) [inline src/klein/_typing.py:13]: Will get rid if this by using the zone plugin for mypy in another PR

**Comment by wsanchez** [comment]
In order to ensure that `KleinRoute` is defined properly, we really should add type hints to `Klein.route`.
Moving to draft until that's done.

File: src/klein/_app.py [hunk 5]
```diff
@@ -39,17 +44,30 @@
...
     KleinSynchronousRenderable, Awaitable[KleinSynchronousRenderable]
 ]
-KleinRoute = Callable[[Any, IRequest, VarArg(Any), KwArg(Any)], KleinRenderable]
-KleinErrorHandler = Callable[
-    [Optional["Klein"], IRequest, Failure], KleinRenderable
-]
+
+
+class KleinRoute(Protocol):
+    def __call__(self, request: IRequest) -> KleinRenderable:
```
Comment (wsanchez) [inline src/klein/_app.py:56]: Note that this removes `*args` and `**kwargs`.

If those a present, then any method decorated with `@route` has to accept them, and that's not currently (and doesn't want to be) a requirement.

File: src/klein/_app.py [hunk 6]
```diff
@@ -268,7 +291,10 @@ def index(request):
 
         @named("router for '" + url + "'")
         def deco(f: KleinRoute) -> KleinRoute:
-            kwargs.setdefault("endpoint", f.__name__)
+            kwargs.setdefault(
+                "endpoint",
+                f.__name__,  # type: ignore[attr-defined]
```
Comment (wsanchez) [inline src/klein/_app.py:296]: To remove the `ignore` here would mean adding the `__name__` attribute to `KleinRoute`.

I think we should probably do that, but I'm holding off on that until I can get the type hints for `Klein.route` to work.

**Comment by wsanchez** [comment]
Currently stuck trying to type `Klein.route` properly:
 * https://mypy-play.net/?gist=b14a18f0e5ac4d2b8805f8653803731d&mypy=latest&python=3.9
 * https://mypy-play.net/?gist=2d025a1e75233dcc5e43a2c99090aafa&mypy=latest&python=3.9&flags=strict

**Review by adiroiban — APPROVED** [review]
Looks good.
While checking this, I found that klein required twisted 16.6 but in tox we still have

```
deps =
    tw160: Twisted==16.0.0
    tw161: Twisted==16.1.1
    tw162: Twisted==16.2.0
    tw163: Twisted==16.3.2
    tw164: Twisted==16.4.1
    tw165: Twisted==16.5.0
```

All good. Thanks!

Look forward to using the zope interface extension :)

**Pull request merged**